src/rust/bitbox02-rust/src/hal.rs

```diff
@@ -58,24 +86,66 @@ impl Hal for BitBox02Hal {
 
 #[cfg(feature = "testing")]
 pub mod testing {
+    use alloc::collections::BTreeMap;
+    use alloc::string::String;
+    use alloc::vec::Vec;
+
     pub struct TestingSd {
         pub inserted: Option<bool>,
+        files: BTreeMap<String, BTreeMap<String, Vec<u8>>>,
     }
 
     impl TestingSd {
         pub fn new() -> Self {
-            Self { inserted: None }
+            Self {
+                inserted: None,
+                files: BTreeMap::new(),
+            }
         }
     }
-    pub struct TestingHal<'a> {
-        pub ui: crate::workflow::testing::TestingWorkflows<'a>,
-        pub sd: TestingSd,
-    }
 
     impl super::Sd for TestingSd {
         fn sdcard_inserted(&mut self) -> bool {
             self.inserted.unwrap()
         }
+
+        fn list_subdir(&mut self, subdir: Option<&str>) -> Result<Vec<String>, ()> {
+            match subdir {
+                Some(key) => Ok(self
+                    .files
+                    .get(key)
+                    .map(|files| files.keys().cloned().collect())
+                    .unwrap_or_default()),
+                None => Ok(self.files.keys().cloned().collect()),
+            }
+        }
+
+        fn erase_file_in_subdir(&mut self, filename: &str, dir: &str) -> Result<(), ()> {
+            self.files
+                .get_mut(dir)
+                .and_then(|files| files.remove(filename).map(|_| ()))
+                .ok_or(())
+        }
+
+        fn load_bin(
+            &mut self,
+            filename: &str,
+            dir: &str,
+        ) -> Result<zeroize::Zeroizing<Vec<u8>>, ()> {
+            self.files
+                .get(dir)
+                .and_then(|files| files.get(filename))
+                .map(|data| zeroize::Zeroizing::new(data.clone()))
+                .ok_or(())
+        }
+
+        fn write_bin(&mut self, filename: &str, dir: &str, data: &[u8]) -> Result<(), ()> {
+            self.files
+                .entry(dir.into())
+                .or_default()
+                .insert(filename.into(), data.to_vec());
+            Ok(())
+        }
     }
 
     impl TestingHal<'_> {
```

Comment:
Isn't it a bit weird to have the `impl` before the `struct`?

Reply:
Moved down one too much :see_no_evil: fixed